### DISCUSSION TITLE
How does Music Assistant fetch the metadata for my music?

### DISCUSSION THREAD

**Discussion by marcelveldt** [music-assistant/support#543] [Show and tell]
If you have local musicfiles (e.g. mp3 or flac files) it is very important that the ID3 tags are correct on your music files. Based on these tags, your music will be loaded into Music Assistant with the correct name, metadata etc.
You can use an ID3 tag editor such as MusicBrainz Picard to correct/set the tags on your music files.
For online music sources (e.g. Spotify) the metadata are retrieved from the music provider(s) itself.

### Basic metadata

- **title**: read from "title" ID3 tag, fallback to filename if the tag is missing.

- **version**: parsed from the title. For example "My great song (live version)" will be parsed as title "My great song" and version "live"

- **track artist(s)**: The Track artist(s) will be read from "artist" ID3 tag. Multiple artists are allowed in that field by a few common separators (e.g. semicolon). If no artist ID3 tag is present, it will try to parse it from the filename (e.g. artist - title.mp3).

- **album title**: The album title will be read from the "album" ID3 tag. A track will NOT be assigned to any album if the album tag is missing. Just like the track title, the version of the album will be parsed from the title.

- **album artist**: The Album artist(s) will be read from the "albumartist" ID3 tag. Just like regular track artist(s), this accepts multiple artists with the common separators. If the albumartist ID3 tag is missing, the whole album will be assigned to "Various Artists.

- **Album type**: If there are 5 tracks or less within an album, the albumtype will be parsed as "single". If most/all tracks of the album have different artists, the album version will be set to "compilation". In all other cases the type will be set to just "album".

- **isrc**: will be read from ID3 tags if present. ISRC is used to match tracks on streaming providers. If no ISRC present we fallback to exact name matching. Having ISRC on the ID3 tracks saves a lot of comparing and thus is faster. I recommend to have it when you have streaming providers and files (but things will work without).

- **disc and track number**: will be read from ID3 tags if present.

- some other metadata will be read too from the ID3 tags if present, such as lyrics, genre(s) and copyright. In fact all tags that are supported by the "Tinytag" python library used under the hood.

### Images and additional metadata

- If the musicfile contains an embedded image, that is preferred as main image and copied over to the album as fallback.

- If the music is stored as "AlbumArtist\Album\song.ext" file structure, we will look for additional metadata there.

- If we did not find additional metadata stored within folders or the source is a streaming provider, we will fetch the additional metadata from online sources such as fanrt.tv and the audiodb. To not stress those providers with useless requests we only fetch the metadata in a lazy manner. Only when view the details of an artist or album or when you play one of the tracks, this metadata will be fetched. This means that when you do not have local metadata you will slowly get this metadata filled into the db.

- If there is an Artist folder found, we check for images there, like fanart(s), logos etc. If there is an artist.nfo file found, we grab the additional metadata from there. [We follow the pattern from Kodi here](https://kodi.wiki/view/NFO_files/Artists)

- If there is an AlbumArtist folder found, we check for images there, like cover(s), discart etc. If there is an album.nfo file found, we grab the additional metadata from there. [We follow the pattern from Kodi here](https://kodi.wiki/view/NFO_files/Albums)

- We do not (yet) store any online fetched metadata into the folders on disk but this might be something for a future release to add (with a toggle to enable it).

### Matching Artist/Album/Song on multiple providers

If you use MA with multiple music providers we match the same items on multiple providers. We do this by strict matching.
For each track we store a reference for each quality. You can have the same song but with different qualities. For example losless FLAC file on disk but compressed ogg vorbis on a streaming provider. When playback is requested of a song, we always sort by quality and availability meaning we prefer the highest quality.

**matching artists**
- prefer MusicBrainz ID match
- fallback to exact name match including compare of albums and tracks to make sure we're not matching an artist with the same name.

**matching albums**
- prefer MusicBrainz ID match or UPC match
- Album title must exactly match
- Album version (e.g. live, remastered etc.) must exactly match
- Album artist(s) must exactly match

**matching tracks**
- prefer MusicBrainz ID match or ISRC match
- Album must exactly match
- Title must exactly match
- Version must exactly match
- Fallback to duration match if both tracks have compilation (multi artist) albums

### Correcting changes

If you've spotted some errors in the metadata from your files, e.g. an album attached to the wrong artist, you can correct the issue with a ID3 tag editor. This change will be picked up by MA on the next sync interval.

#### Sync of library

MA syncs the library items of the streaming providers and the files on disk to the database. The first time this sync runs this takes a lot of time (depending on hardware and collection size) because MA has to read all tags from the files which is slow. This is one time only. Once the first sync is completed, MA switches to delta syncs which only take seconds. If teh initial sync gets aborted somehow by accident, the sync will continue where it left off.

The MA (delta) sync runs every 3 hours. This is a very fast sync because only items with a changed checksum are processed. For the filesystem provider this means that the timestamp changed. The delta sync 

You can trigger a manual sync by simply reloading the integration or HA. Besides that you can trigger a manual sync within the library view (e.g. Artists) and click the menu button on the top right (the 3 dots).

It is not possible to force a full sync / rescan but that should not be needed in normal circumstances.
If things are really wrong and you need to start over, you can remove the integration and re-add it, which will wipe the database. Only do that as a last resort because you will loose all metadata in the database including playcounts and normalization data.

**Comment by OzGav** [comment]
Thanks for that detailed info @marcelveldt I would just like to show this example where it does not seem to align with what you have written above about matching albums. This isn't a problem for me as I am not seeing lots of duplicate entries but I just wanted to highlight it in case it is not your desired behaviour. In this first image I have selected ARTISTS>>ABBA. I only have one CD on my NAS which is called The Definitive Collection https://musicbrainz.org/release-group/e2996feb-86b9-3550-88ec-f1c9c9af3961 As far as I can tell this Album is NOT on Spotify. However you can see some but not all tracks have matched:
![image](https://user-images.githubusercontent.com/19848947/176330375-47bc43c4-dad5-4f48-bd7f-7406bcc6d69b.png)
Scrolling down a bit further we can see not a match and I wonder if it is because of the case of the word "on". The other thing I found interesting here is the Spotify entry is from the Album "Super Trouper" but no other tracks from that Album are shown
![image](https://user-images.githubusercontent.com/19848947/176330685-eb3275f8-8415-4af8-a376-74567c9868e0.png)
This next snippet shows another pair of entries similar to the last but a different Spotify album
![image](https://user-images.githubusercontent.com/19848947/176330800-9246fd76-8a48-4c0e-879d-6e6b79d07480.png)
Is all of this what you expect to happen?

**Reply by marcelveldt** [reply]
This is possible because the same tracks (as in same recording with same isrc or MusicBrainzRecordingId) can be used on different albums. If you would open the context menu for one of those tracks and go the details, I'm pretty sure you will find that this is exactly the same track (you can verify with the preview buttons).

So a track can belong to 1 or more albums) In this case that is pretty obvious because "The Definitive Collection" is a Compilation. It contains tracks from previous albums.

It can also happen for various artists Collections, such as "Best Hits of the 80s". In that case such an album contains tracks from many artists (and albums) and the track linking will work too in that case. It is exactly the same as Spotify's own track linking for tracks within playlists.

Does that make sense?

**Comment by OzGav** [comment]
Yes it does and I’m happy with that as it removes duplicates. I guess I am just wondering if it is an issue that some but not all tracks matched? And whether those two entries that look like a case mismatch is worth fixing or able to be fixed!

**Reply by marcelveldt** [reply]
It is all "lazy" so maybe just click one of the tracks and see if it matches ?
You can use the "refresh info" button from the details contextmenu to re-trigger the matcher. This will force an update of the item from all providers.

![Schermafbeelding 2022-06-29 om 09 07 43](https://user-images.githubusercontent.com/6389780/176373757-da06518d-7f3c-4ddb-8368-a7af3eef83b2.png)

**Comment by OzGav** [comment]
I’ve seen that before but now I am seeing SHOW INFO not REFRESH ITEM?

Edit. Actually found it top right. So that seems to refresh the artist. Don’t think I can just refresh a track

**Reply by marcelveldt** [reply]
Open the details for a track. Track --> contextmenu on listitem --> show info --> contextmenu top right --> refresh info

**Reply by marcelveldt** [reply]
Each media type has its own details view. So you can view the full info for an artist, album and even track.
In reality you will probably never view the details of a track but it's there :-)

**Reply by OzGav** [reply]
Yes cool I can see that refresh is doing something but no change to the other entries….

**Reply by marcelveldt** [reply]
What exactly is the issue that you still see ? Do note that it possible that there are no matches on Spotify for the isrc's of your local files.

**Reply by OzGav** [reply]
Mainly just wanted to make sure you were happy this was working correctly. Maybe the two that didn’t match on case (if that indeed is the problem)?

**Reply by marcelveldt** [reply]
Yesterday I fixed that "case issue" but I'm not sure it will be corrected if it is already wrong in the db.
I think the behavior you see is correct, especially with compilation albums like this. I would be more surprised if some tracks have no match if there was an exact match for an album found on both providers

**Reply by OzGav** [reply]
Yes I just can’t tell how “exact” the match is. For their most popular songs when there are a lot of options in Spotify it is matching but other with less than 10 options it isn’t. See below for an example. Can I do any refreshing anywhere to see if that case issue resolves?
![E0A4A750-B11C-488A-AC5E-831F412EE60B](https://user-images.githubusercontent.com/19848947/176403475-493c1b85-0cbf-4420-9d4f-1b2509e5b4ab.jpeg)

**Comment by arpel** [comment]
Hello,
My library is exclusively made of FLAC files which does not seem to properly have ID3 tags (hence everything falling into "Various Artists").
FLAC does not officially support ID3, is there a way to use folder/file naming and/or external sources (as Jellyfin seems to do) to get everything properly sorted ?

**Reply by OzGav** [reply]
Yes FLAC does. Tag your files with Picard and you will be all set

**Reply by arpel** [reply]
You're right and that's exactly what I did. There are tons of contradictory information out there : your approach seems to be the most consistent one (relying only on tags) but I admit that the "self management" of Jellyfin was easier to start with.

**Comment by Lush78** [comment]
My YouTube music doesn't display the metadata or artwork. My entire flac library is what I've uploaded and it has properly embedded metadata and art that is recognized in the YouTube interface fine. When playing through MA is only shows the song title on my Google displays. How can I correct this?

**Reply by OzGav** [reply]
@MarvinSchenkel may  know what YT is doing but if you have your own FLAC library then you will have a way better experience if you just share that local library directly with MA via the file system provider.

**Reply by Lush78** [reply]
That's my end goal here, but I'm currently unable to get large playlists to work as you seen in my other thread. I do want to eventually get my playlists working with my hi-res flac files. I'm just not sure how to get there at the moment.

**Reply by OzGav** [reply]
You can still share your library locally. MA will then link to the YouTube tracks. Your playlists will be visible but when the tracks are played the local tracks should be used.

**Reply by MarvinSchenkel** [reply]
Smells like another inconsistency in YTM. This would require a fair amount of work to support 'personally uploaded songs to YTM' and to be honest it's not something that is high on my priority list. Like @OzGav mentioned, the local file system provider is the way to go for your flac library. I would say, please hang on until there's a fix for your large playlist

**Reply by Lush78** [reply]
What is strange, is the first song displays the metadata fine on the google hub devices (except album art).  The following songs all report the song title properly, but the artist is always "Music Library Uploads".  Yet in the music assistant queue, it displays all the proper metadata (artist, album, song title) including album art.

**Reply by Lush78** [reply]
> Smells like another inconsistency in YTM. This would require a fair amount of work to support 'personally uploaded songs to YTM' and to be honest it's not something that is high on my priority list. Like @OzGav mentioned, the local file system provider is the way to go for your flac library. I would say, please hang on until there's a fix for your large playlist

Am I waiting for something on the music assistant end for the large playlist support?  The same playlists work fine on my Android based sony walkman that i use for playing the same flac/hi-res files in my work office.  I feel like my home assistant OS running directly and alone on my slightly older i5/8gb dell optiplex should be capable hardware for this function.

**Reply by OzGav** [reply]
As I said in your other question if MA is failing to load your playlist beyond 800 songs then open an issue. There are people with playlists bigger than that where it works fine so look at what is maybe causing a problem. Marvin  said above that “personally uploaded songs” are not currently supported. If that is the problem then open a feature request.

**Comment by dutchmendesign** [comment]
Hello Marcel,
First of all i really appreciate your works! It made me rediscover my 15y old mp3 music collection :-)
Now I am re-tagging the folders.

I have a question about duo released albums.
What is the best approach for MA?

I tried to implement the 'albumartist' tag as : "duo name; artist1; artist2"
For example: https://musicbrainz.org/release/7c83ee2d-b977-4211-a9f2-6721fc2ba4f4

I have 3 different artist folders on the share: "Miss Kittin", "Golden Boy", "Golden Boy with Miss Kittin".
In these folders i can put artist metadata; i started with just 'folder.jpg' and 'fanart.jpg'.

For the folder "Golden Boy with Miss Kittin / Or" i can put the album metadata.

In Picard I change tags like below:
![Schermafbeelding 2024-07-17 160948](https://github.com/user-attachments/assets/047054ea-fb7e-4f6b-b835-9c20c3bbff64)


I already tried several solutions but in the end some metadata gets messed up.
For example in MA i read the artist1 biography which should be artist2 biography, and artis1 biography is disappearing.
Also songs from other albums are getting matched to artist1 instead of artist2, while not making any changes to these id3 tags.
I am guessing there is a connection with musicbrainz_artistid and musicbrainz_albumartistid but i don't fully understand yet.
Tried "musicbrainz_artistid: 'null; 9663a4dd-7afa-4314-a4f3-b444bdedcdda; af6bd973-3e30-4339-94b5-85b139362e6d'" but it didnt work either.

Any suggetions?
Thank you in advance!

**Reply by OzGav** [reply]
This seems to be a unique case. Golden Boy is likely the main artist so I would put this album in a folder “Golden Boy” with their artist thumb and fanart. Then I would create a folder for “Miss Kittin” and just put the artist thumb and fanart in there.

MA should link the tracks to both artists and should pull in the artwork as well.

tag the files with Picard and don’t make any manual changes.

**Reply by dutchmendesign** [reply]
Hi OzGav, your reply is appreciated.
I would like to have the album listed as it is originally issued AND for both artists.
For example, as i managed for 'First Album' by 'Miss Kittin & The Hacker'.
The tracks are also listed by the individual artists as a by-catch; it's nice but not really necessary (to me).

![Schermafbeelding 2024-07-17 201118](https://github.com/user-attachments/assets/55f8bd35-54c4-4489-abcd-a5cf2fe51716)

![Schermafbeelding 2024-07-17 201143](https://github.com/user-attachments/assets/f666a3ae-2e39-49ae-9785-ddd5f731be26)

![Schermafbeelding 2024-07-17 202552](https://github.com/user-attachments/assets/c8b4a589-4629-496a-9325-a33bdf3278ac)

![Schermafbeelding 2024-07-17 203951](https://github.com/user-attachments/assets/3dc0d80c-e0e9-4cb0-a117-a95de52e6a1b)

The challenge might be that 'Golden Boy with Miss Kittin' is not recognized as a group in MusicBrainz, while  'Miss Kittin & The Hacker' is.

The tracks are not linked to both artists (Albumartists: Artist1; Artist2) when i keep the original Picard data as I tried for several other albums. As far as i have experienced i can only find Artist1 in the MA DB. For example https://musicbrainz.org/release/7d4a86ea-b90a-4424-9348-9e305944c9ed. It is not listed with 'Beth Gibbons & Rustin Man' or 'Rustin Man' as i would expect, only 'Beth Gibbons'.

**Reply by OzGav** [reply]
> The challenge might be that 'Golden Boy with Miss Kittin' is not recognized as a group in MusicBrainz, while 'Miss Kittin & The Hacker' is.

Correct. MA can only go off the supplied tags. If you have an example where there are multiple artists in the tags but the track is not shown against all of those artists then that can be looked at.

In you screenshots above that is likely a bug which has been fixed in rc2, an artist with a & in the name should not have been split apart.

**Reply by dutchmendesign** [reply]
Okay, great! I will wait for the update then. Thanks!

An additional example where only 1 artist is recognized.
Same applies to "Artist 1 'with' 'Artist 2".

![Schermafbeelding 2024-07-18 091946](https://github.com/user-attachments/assets/77bc6623-b26a-4d33-8719-7632b55ddc5a)

![Schermafbeelding 2024-07-18 093842](https://github.com/user-attachments/assets/84fc4cc9-4105-4359-93c5-34fae9c02711)

![Schermafbeelding 2024-07-18 095308](https://github.com/user-attachments/assets/3e3fb287-41cc-4770-beb8-2fa5b4950ce9)

**Reply by OzGav** [reply]
It should be listed under Ellen Allien as well which it looks like it might be (based on the first image)

**Reply by OzGav** [reply]
If it’s not it might be because of the problem I mentioned. Look in the logs for errrors

**Reply by dutchmendesign** [reply]
> It should be listed under Ellen Allien as well which it looks like it might be (based on the first image)

Yes, it is only listed under 1 artist, in this case Ellen Allien.

**Reply by dutchmendesign** [reply]
Although there is the strange behavior that some tracks seem to be recognized for both artists e.g. Track 11 'Rippin Kittin'.

![Schermafbeelding 2024-07-18 104440](https://github.com/user-attachments/assets/18091153-a7ee-4435-8ce9-31a81ba58966)

In this case the artists 'Golden Boy' and 'Miss Kittin' are individually listed for this track.

![Schermafbeelding 2024-07-18 114117](https://github.com/user-attachments/assets/9ed454b4-6de3-4b8c-8267-b87e91fba343)

For another album (Or) they are not both listed, just 'Golden Boy' for album and tracks.
https://musicbrainz.org/release/bd8a3a12-2aa4-421c-888a-a026d8ad97b1

**Reply by OzGav** [reply]
You will need to do a careful examination of the tags. If they are identical but MA is behaving inconsistently then open an issue.
